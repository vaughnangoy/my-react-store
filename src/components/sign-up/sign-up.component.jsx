import React from 'react';
import './sign-up.styles.scss';

import FormInput from '../form-input/form-input.component';
import CustomButton from '../customButton/customButton.component';

import { auth, createUserProfile } from '../../config/firebase-config';

export class SignUp extends React.Component {
    constructor() {
        super();

        this.state = {
            displayName: '',
            email: '',
            password: '',
            confirmPassword: '',
        };
    }

    handleThisSubmit = async e => {
        e.preventDefault();
        const { displayName, email, password, confirmPassword } = this.state;
        this.setState({ errorMessage: null });

        if (password !== confirmPassword) {
            this.setState({ errorMessage: 'Passwords must be the same' });
            return;
        }

        try {
            const { user } = await auth.createUserWithEmailAndPassword(
                email,
                password,
            );
            await createUserProfile(user, { displayName });
            this.setState({
                displayName: '',
                email: '',
                password: '',
                confirmPassword: '',
            });
        } catch (e) {
            console.log(e);
            this.setState({ errorMessage: e.message });
        }
    };

    handleChange = e => {
        const { value, name } = e.target;
        this.setState({ [name]: value });
    };

    render() {
        return (
            <div className="sign-up">
                <h2 className="title">I do not have an account</h2>
                <span>Sign up with your email and password</span>
                {this.state.errorMessage ? (
                    <div className="error">
                        <em>{this.state.errorMessage}</em>
                    </div>
                ) : (
                    ''
                )}
                <form className="sign-up-form" onSubmit={this.handleSubmit}>
                    <FormInput
                        type="displayName"
                        name="displayName"
                        id="displayName"
                        value={this.state.displayName}
                        required="required"
                        handleChange={this.handleChange}
                        label="displayName"
                        to="displayName"
                    />
                    <FormInput
                        type="email"
                        name="email"
                        id="email"
                        value={this.state.email}
                        required="required"
                        handleChange={this.handleChange}
                        label="email"
                        to="email"
                    />
                    <FormInput
                        type="password"
                        name="password"
                        id="password"
                        value={this.state.password}
                        required="required"
                        handleChange={this.handleChange}
                        label="password"
                        to="password"
                    />
                    <FormInput
                        type="password"
                        name="confirmPassword"
                        id="confirmPassword"
                        value={this.state.confirmPassword}
                        required="required"
                        handleChange={this.handleChange}
                        label="confirmPassword"
                        to="confirmPassword"
                    />
                    <CustomButton type="submit" data-testid="sign-up-form">
                        Sign Up
                    </CustomButton>
                </form>
            </div>
        );
    }
}
