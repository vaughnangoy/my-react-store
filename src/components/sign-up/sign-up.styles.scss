.sign-up {
    display: flex;
    flex-direction: column;
    width: 380px;

    .title {
        margin: 10px 0;
    }

    .error {
        color: #990000;
        font-weight: 600;
        padding: 5px 15px;
        margin: 20px 0;
        background-color: #fab6b6;
        border-radius: 5px;
    }
}
