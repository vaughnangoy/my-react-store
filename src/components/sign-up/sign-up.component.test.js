import React from 'react';
import { cleanup, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { SignUp } from './sign-up.component';

describe('SignUp', () => {
    afterEach(cleanup);
    it('renders', () => {
        const { asFragment } = render(<SignUp />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should fill in and submit the form', () => {
        const handlerSpy = jest.fn();
        const { getByLabelText, getByText, getByTestId } = render(
            <SignUp handleSubmit={handlerSpy} />,
        );
        fireEvent.change(getByLabelText('displayName'), {
            target: { value: 'Test' },
        });
        fireEvent.change(getByLabelText('email'), {
            target: { value: 'Test@test.com' },
        });
        fireEvent.change(getByLabelText('password'), {
            target: { value: 'test123' },
        });
        fireEvent.change(getByLabelText('confirmPassword'), {
            target: { value: 'test123' },
        });
        fireEvent.click(getByTestId('sign-up-form'));
    });
});
