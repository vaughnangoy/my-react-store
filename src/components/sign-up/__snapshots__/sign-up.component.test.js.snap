// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp renders 1`] = `
<DocumentFragment>
  <div
    class="sign-up"
  >
    <h2
      class="title"
    >
      I do not have an account
    </h2>
    <span>
      Sign up with your email and password
    </span>
    <form
      class="sign-up-form"
      data-testid="sign-up-form"
    >
      <div
        class="group"
      >
        <input
          class="form-input"
          id="displayName"
          name="displayName"
          required=""
          type="displayName"
          value=""
        />
        <label
          class=" form-input-label"
          for="displayName"
        >
          displayName
        </label>
      </div>
      <div
        class="group"
      >
        <input
          class="form-input"
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
        <label
          class=" form-input-label"
          for="email"
        >
          email
        </label>
      </div>
      <div
        class="group"
      >
        <input
          class="form-input"
          id="password"
          name="password"
          required=""
          type="password"
          value=""
        />
        <label
          class=" form-input-label"
          for="password"
        >
          password
        </label>
      </div>
      <div
        class="group"
      >
        <input
          class="form-input"
          id="confirmPassword"
          name="confirmPassword"
          required=""
          type="password"
          value=""
        />
        <label
          class=" form-input-label"
          for="confirmPassword"
        >
          confirmPassword
        </label>
      </div>
      <button
        class=" custom-button"
        type="submit"
      >
        Sign Up
      </button>
    </form>
  </div>
</DocumentFragment>
`;
